docs/sdk-configuration.md

```diff
@@ -0,0 +1,492 @@
+    tracer.spanBuilder("export").startSpan();
+  }
+}
+public class OpenTelemetrySdkBuilder {
+  public OpenTelemetrySdkBuilder addSpanExporter(SpanExporter exporter) {
+    tracerProvider.addSpanExporter(exporter);
+    components.add(exporter);
+  }
+
+  public OpenTelemetrySdkBuilder setSampler(Sampler sampler) {
+    tracerProvider.setSampler(sampler);
+    components.add(sampler);
+  }
+
+  public OpenTelemetry build() {
+    OpenTelemetrySdk sdk = new OpenTelemetrySdk(tracerProvider.build(), meterProvider.build());
+    for (OpenTelemetryComponent component : components) {
+      component.setOpenTelemetry(sdk);
+    }
+  }
+}
+```
+
+A framework author will have an even easier time since most dependency injection frameworks
+natively support lazy injection.
+
+```java
+@Component
+public class MonitoringModule {
+  
+  @Bean
+  @ForSpanExporter
+  public Tracer tracer(TracerProvider tracerProvider) {
+    return tracerProvider.get("spanexporter");
+  }
+}
+
+@Component
+public class MyExporter implements SpanExporter {
+
+  private Lazy<Tracer> tracer;
+  
+  @Inject
+  public MyExporter(@ForSpanExporter Lazy<Tracer> tracer) {
+    this.tracer = tracer;
+  }
+  
+  @Override
+  public void export() {
+    tracer.get().spanBuilder("export").startSpan();
+  }
+}
+```
+
+## Immutability of OpenTelemetry configuration
+
+The above attempts to avoid allowing a built SDK to be mutated. Allowing mutation can make code
+harder to reason about (any component, even deep in business logic, could update the SDK without
+hindrance), can reduce performance (require volatile read on most operations), and produce thread
+safety issues if not well implemented. In particular, compared to the current state as of writing,
+
+- `addSpanProcessor` is not needed because we instead push the complexity of handling telemetry
+within telemetry components to those components, where the maintainers will have more domain
+knowledge. It allows this mutator method to be removed from the end-user API.
+
+- `updateTraceConfig` - instead of allowing updates at the top level, we should consider making
+`TraceConfig` an interface, and the SDK default implementation is static. It allows the above
+benefits of immutability to be in place for the common case where dynamic updates are not needed.
+Where dynamic updates are needed, it can be replaced with a mutable implementation instead of making
+the SDK configuration mutable. This keeps update methods out of the end-user APIs and will generally
+give framework developers more control by handling dynamicism themselves without the chance of
+end-users to affect it negatively.
+
+Some highly buggy code that could be enabled by mutability.
+
+```java
+
+class SleuthUsingService {
+  
+  @Inject
+  private OpenTelemetry openTelemetry;
+
+  public void doLogic() {
+    // My logic is important, so always sample it!
+    OpenTelemetrySdk.getTracerManagement().updateTraceConfig(config -> config.setSampler(ALWAYS_ON));
```

Comment:
Who is the author of this line? Only one person should configure or control the SDK config. Any random "service" running in an application service should be able to affect only their instance of SDK if any, otherwise should not be able to control the config.

Reply:
It seems like this might be an argument against exposing a global where anyone could muck about with it and effect everything running in the current classloader (or JVM itself depending on where the global is class loaded first).